src/typechecker/FStar.TypeChecker.Env.fs

```diff
@@ -175,10 +175,11 @@ and env = {
   nosynth        :bool;                         (* don't run synth tactics *)
   uvar_subtyping :bool;
   tc_term        :env -> term -> term*lcomp*guard_t; (* a callback to the type-checker; g |- e : M t wp *)
-  type_of        :env -> term -> term*typ*guard_t;   (* a callback to the type-checker; g |- e : Tot t *)
-  type_of_well_typed :env -> term -> typ;          (* a callback to the type-checker; falls back on type_of if the term is not well-typed *)
+  typeof_tot_or_gtot_term:env -> term -> bool -> term*typ*guard_t;   (* a callback to the type-checker; g |- e : Tot t *)
+  typeof_tot_or_gtot_term_fastpath  :env -> term -> option<typ>;       (* a callback to the type-checker, uses fast path *)
   universe_of    :env -> term -> universe;           (* a callback to the type-checker; g |- e : Tot (Type u) *)
-  check_type_of  :bool -> env -> term -> typ -> guard_t;
+  tc_check_tot_or_gtot_term_maybe_fastpath :env -> term -> typ -> bool -> bool -> guard_t;
+  universeof_fastpath    :env -> term -> option<universe>;
```

Comment:
Would be good to update the comments here. And to indicate there why we are exposing three versions of the function in the interface, i.e., typeof_, typeof_..._fastpath, tc_check_..._maybe_fastpath

Reply:
likewise for universeof

Reply:
I will add a comment. But here is my reading of it:

* The `tc_` functions have a return type of the form `env -> term -> term & lcomp & guard_t`
* The `tc_check_` functions in addition take an expected type as argument: `env -> term -> typ -> term & lcomp & guard_t`
* The `typeof_` functions return only a type not complete lcomp: `env -> term -> term & typ & guard_t`
* Then there are `_fastpath` variants, which I added, that take the fastpath (and return no guard for example).
* The `maybe_fastpath` variant first tries the fastpath, and if it fails, tries the slow path.